src/experiments/logins/api.js

```diff
@@ -0,0 +1,164 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/* globals AppConstants, browser, Components, CustomizableUI, dispatcher,
+   ExtensionCommon, ExtensionAPI, Services, XPCOMUtils */
+
+"use strict";
+
+ChromeUtils.defineModuleGetter(this, "ExtensionUtils",
+                               "resource://gre/modules/ExtensionUtils.jsm");
+ChromeUtils.defineModuleGetter(this, "LoginHelper",
+                               "resource://gre/modules/LoginHelper.jsm");
+ChromeUtils.defineModuleGetter(this, "Services",
+                               "resource://gre/modules/Services.jsm");
+
+const { ExtensionError } = ExtensionUtils;
+
+const getLogins = () => {
+  const logins = Services.logins.
+    getAllLogins().
+    filter((l) => !(l.hostname || "").startsWith("chrome://")).
+    map(LoginHelper.loginToVanillaObject);
+  return logins;
+};
+
+const getLogin = (id) => {
+  const login = getLogins().
+    filter(l => l.guid === id)[0];
+  return login || null;
+};
+
+this.logins = class extends ExtensionAPI {
+  getAPI(context) {
+    const EventManager = ExtensionCommon.EventManager;
+    return {
+      experiments: {
+        logins: {
+          // See schema.json for function documentation.
+          getAll() {
+            const logins = getLogins();
+            return logins;
+          },
+          get(id) {
+            const login = getLogin(id);
+            return login;
+          },
+          // TODO: should this be called create(), or should the event name be changed to onAdded?
+          add(loginInfo) {
+            if (getLogin(loginInfo.guid)) {
+              throw new ExtensionError(`Add failed: Login already exists with GUID ${loginInfo.guid}`);
+            }
+            try {
+              const login = LoginHelper.vanillaObjectToLogin(loginInfo);
+              Services.logins.addLogin(login);
+            } catch (ex) {
+              throw new ExtensionError(ex);
+            }
+            const createdLogin = getLogin(loginInfo.guid);
+            return createdLogin;
+          },
+          update(loginInfo) {
+            const login = getLogin(loginInfo.guid);
+            if (!login) {
+              throw new ExtensionError(`Update failed: Login not found with GUID ${loginInfo.guid}`);
+            }
+            const loginAndMetaData = LoginHelper.newPropertyBag(loginInfo);
+            Services.logins.modifyLogin(login, loginAndMetaData);
+            const updatedLogin = getLogin(loginInfo.guid);
+            return updatedLogin;
+          },
+          remove(id) {
+            const login = getLogin(id);
+            if (!login) {
+              throw new ExtensionError(`Remove failed: Login not found with GUID ${id}`);
+            }
+            try {
+              Services.logins.removeLogin(LoginHelper.vanillaObjectToLogin(login));
+            } catch (ex) {
+              throw new ExtensionError(ex);
```

Comment:
Error handling is still a little bit WIP here. If we don't rethrow errors as `ExtensionError`s, then the promise is still rejected, but the original error info is destroyed, and the webextensions code emits a particularly foul "An unexpected error occurred" message...but maybe we shouldn't expose all the errors in this way. Feedback welcome :-)

Reply:
With everything else declared in the schema, I'm almost surprised we don't have to specify the inventory of exceptions that could be thrown here rather than being able to clone-by-rethrow here. That's really the only thing I could think of to do differently - i.e. declare all the expected exceptions. But that feels possibly a little too boilerplatey

Reply:
It might require changes to the underlying nsILoginManager implementations to get better enumerations right now.

Eventually worth it, but for this initial work.

Reply:
@lmorchard Yeah, the webextensions code in firefox doesn't validate or even parse the return values at all. I was told I could specify return types, but was afraid to do so: it's so easy to typo an unvalidated schema, and if they add return type checking, it might well break the API in the wild. I'm choosing to look at it as a "schema is half-full" situation ;-P